### ISSUE TITLE
Fix LED flickering issue

### ISSUE THREAD

**Issue by Elioty** [Elioty/ledWall#10]
Tracks:
- https://github.com/adafruit/Adafruit_TLC59711/pull/4
- https://forums.adafruit.com/viewtopic.php?f=47&t=48075&sid=4c119cdc31bfa8921568c2ed913ed6bf
- https://github.com/ulrichstern/Tlc59711/wiki
- https://forum.pjrc.com/threads/28460-teensy3-1-tlc59711-16-bit-led-driver-timing-issue-and-flicker

**Comment by WolfWings** [comment]
As a note, with further research I found that this chip is INCREDIBLY sensitive to timing, but if you have a buffered SPI transmission method you can actually go all the way up to (CPU / 2)Mhz update rate without issue on most AVR Ardunio's via MSPIM mode for example, and with Teensy's I was able to go up to the datasheet-rated 10Mhz without issue.

The main thing is that any gaps in the SPI stream between bytes cause a build-up of buffering inside the chips, which causes the failed clocking and false propagation. So the double-buffered MSPIM mode or other similarly 'fully stable' SPI signalling works fine.

**Comment by Elioty** [comment]
Wow, I wasn't expecting anyone to come in here as my project is still in a rather very experimental state.

But yup, I agree this chip is really sensitive to timing. And it still suffers a little bit of flickering from time to time with the current software. But as you can see in [/software/embed/drivers/led_driver.cpp](https://github.com/Elioty/ledWall/blob/67b0a8f22917f59d47afa86bfa00b99a7dda8143/software/embed/drivers/led_driver.cpp), I still have two loops (and some shift operations on a 32-bit and a 16-bit value, not good for a poor 8-bit MCU) and it might create little timing deviation in data transmission here and there. I didn't take a look in the assembly code and calculate all the instruction timing though, the problem might come up crystal clear.
Anyway, I think I should try (and would be easier) with using only one data buffer for all the chips so all I have to do in LED_Write is to literally loop and write onto the SPI with all other computations already done.

If there is still an issue, I can try the MSPIM mode (hopefully I am still using prototyping board and haven't done any KiCad work yet), thanks for pointing this out. I knew the AVR's UART hardware had a buffer built-in but didn't notice the hardware used it for its SPI mode.

**Comment by WolfWings** [comment]
Oh, if you're using AVR chips, trust me, stop using the hardware SPI entirely unless you need to run the AVR in slave mode, switch to MSPIM entirely ASAP and you'll be much happier. You get a two-stage buffer to send SPI bits with, so you can load 8 bits... wait for an interrupt or busy-wait on the status byte saying that byte has been read, and load the next byte while the first byte is still transmitting. So loop overhead becomes a non-issue.

http://imgur.com/qtVzcr6 shows what I pulled off with a single Trinket Pro 3V (so 12Mhz) with MSPIM, updating at the full 6Mhz @ 50fps so there's plenty of leftover CPU time to calculate all those animations in realtime including the EQ effect. :)

And blame me noticing your comment on that pull request ages back to AdaFruit. XD I figured you cared but AdaFruit doesn't so I might as well pass info along to those that'll actually USE it. :)

https://feilipu.me/2015/02/17/avr-atmega-usart-spi-mspim/ is the best document I know of for the MSPIM mode and shows 'scoped waveforms for clarity.

**Comment by Elioty** [comment]
I have used AVR's GPIO, timer (for interrupt and for PWM output to control (servo)motors), ADC, EEPROM and UART hardware for years but I must admit it's certainly the first time I use its SPI so thanks for the advice, I will just switch to MSPIM entirely when I programme SPI master devices. The hardware buffering would indeed allow some loop overhead.

Wow, very nice!! I am still quite far from reaching that level of achievement with my own project haha. I should take some pictures too. And also gotta work on the LED cabling, messy pile of wires at the moment. It's still a prototype though and I still don't know what is the final LED layout I want.

Wow (again), thank you man, let's keep the people power flowing! And long life to FLOSS!
Found that pull request you mentioned and oh well, yea, it's yours and I commented it. Thank you again 😉 

Thanks for that link too but the atmega2560 datasheet is like my bible. It's of course only factual though and doesn't have all the expert tips featured in like "don't bother with hardware SPI in master mode, switch to MSPIM". I will still give it a look, I don't have an oscilloscope at home.

Cheers man

**Comment by hoffmannjan** [comment]
should be fixed in #13
